cupyx/scipy/sparse/linalg/_eigen.py

```diff
@@ -129,15 +129,32 @@ def eigsh(a, k=6, *, which='LM', ncv=None, maxiter=None, tol=0,
     else:
         return cupy.sort(w)
 
+def inversion_eps(dt):
+  '''
+  Inversion epsilon for type `dt`. 
+  Adopted from scipy.sparse.linalg.svds.
+  '''
+  t = dt.char.lower()
+  factor = {'f': 1E3, 'd': 1E6}
+  return factor[t] * cupy.finfo(t).eps
 
 def _lanczos_asis(a, V, u, alpha, beta, i_start, i_end):
+    beta_eps = inversion_eps(a.dtype)
     for i in range(i_start, i_end):
         u[...] = a @ V[i]
         cublas.dotc(V[i], u, out=alpha[i])
         u -= u.T @ V[:i+1].conj().T @ V[:i+1]
         cublas.nrm2(u, out=beta[i])
         if i >= i_end - 1:
             break
+        
+        if beta[i] < beta_eps:
+            V[i+1:i_end,:] = 0
```

Comment:
Beta's are always non-negative, because these are 2-norms of vectors. So there is no point to continue, because all subsequent betas should be even smaller. Hence it is safe to set further eigenvectors to zero.   `which` keyword influence only on `_eigsh_solve_ritz`.

Reply:
I'm sorry that the `which` arg seems irrelevant to the issue.

Anyway, `v.T @ v` has to be close to the identity matrix.

```python
import numpy
import scipy.sparse.linalg
import cupy
import cupyx.scipy.sparse.linalg

for (xp, scp) in [
    (numpy, scipy),
    (cupy, cupyx.scipy),
]:
    w, v = scp.sparse.linalg.eigsh(xp.zeros((4, 4)), k=3)
    print(f'''\
{xp.__name__ = }
{w = }
{v =
}
{v.T @ v =
}
''')
```

results

```
xp.__name__ = 'numpy'
w = array([0., 0., 0.])
v =
array([[ 0.42863411,  0.2627032 ,  0.81395245],
       [ 0.87456566, -0.33167313, -0.35373356],
       [ 0.22335956,  0.69262722, -0.09842178],
       [-0.03896417, -0.58416393,  0.45018566]])
v.T @ v =
array([[ 1.00000000e+00, -1.61633310e-16,  1.99890516e-17],
       [-1.61633310e-16,  1.00000000e+00,  2.13188985e-17],
       [ 1.99890516e-17,  2.13188985e-17,  1.00000000e+00]])

xp.__name__ = 'cupy'
w = array([0., 0., 0.])
v =
array([[0.04644695, 0.        , 0.        ],
       [0.03379258, 0.        , 0.        ],
       [0.17392377, 0.        , 0.        ],
       [0.98308253, 0.        , 0.        ]])
v.T @ v =
array([[1., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])
```

Reply:
Well, while it is possible to continue with some vector `V[i+1]`that is orthogonal to `V[:i+1]` it is a waste of cycles. Because these vectors will be essentially random. Columns of v, that correspond to zero eigenvalues are undefined. Do we really need to mimic  SciPy's  behavior in such details? Any code that relies on it is buggy in fact.

If you insist, I will replace `V[i+1:i_end, :] = 0` by code that generates orthogonal vectors in some quasi-random manner.

If it would be a newly created code I would return truncated `v` and `w` with dimensions, that correspond to actual matrix rank. But as we try to keep consistency with SciPy it is not an option.